update documentations around test isolation (#4797)
cypress-io/cypress-documentation: content/guides/core-concepts/writing-and-organizing-tests.md

<Alert type="success">

<Icon name="check-circle" color="green"></Icon> **Best Practice:** Tests should

> always be able to be run independently from one another **and still pass**.

Reviewer: I think this captures the idempotency talked about in https://github.com/cypress-io/cypress/pull/24316#discussion_r1002060664

</Alert>

> As stated in our mission, we hold ourselves accountable to champion a testing
> process that actually works, and have built Cypress to guide developer towards
> write independent tests from the start.

Reviewer: ```suggestion
As stated in our mission, we hold ourselves accountable to champion a testing
process that actually works, and have built Cypress to guide developers towards
writing independent tests from the start.
```

We do this by cleaning up state _before_ each test to ensure that the operation
of one test does not affect another test later on. The goal for each test should
be to **reliably pass** whether run in isolation or consecutively with other
tests. Having tests that depend on the state of an earlier test can potentially
cause nondeterministic test failures which makes debugging challenging.

Cypress will start each test with a clean test slate by restoring and clearing
all:

- [aliases](/api/commands/as)
- [clock mocks](/api/commands/clock)
- [intercepts](/api/commands/intercept)
- [routes](/api/commands/route)
- [spies](/api/commands/spy)
- [stubs](/api/commands/stub)
- [viewport changes](/api/commands/viewport)

> In additional to a clean test slate, Cypress also believes in run tests in a

Reviewer: ```suggestion
In additional to a clean test slate, Cypress also believes in running tests in a
```

clean browser context such that the application or component under test behaves
consistently when ran. This concept is described as `testIsolation`.

The test isolation mode is a global configuration and can be overridden at the
`describe` level with the
[`testIsolation`](/guides/references/configuration#Global) option.

#### End-to-end testing

Cypress supports the following modes of test isolation in end-to-end testing to
describe if a suite of tests should run in a clean browser context or not:

> `legacy` (deprecated), `on` (experimental) and `off` (experimental).

Reviewer: It feels weird that the only options for test isolation are either `experimental` or `deprecated`. I get it, but it also feels like we don't officially support any of these. Wonder if there is a different way to brand this in the docs?

Author: I could use (current) but it'll be removed/changed in the next few months & wanted to call it out now. Thoughts on `current`?

Reviewer: I am thinking `current` works until we release 11.0, in which we say it's removed. It sounds like we don't actually plan to deprecate it, just remove it.

Reviewer 2: Is there any way we can just scrub legacy from a discussion point? I doubt very many people have even set it. Could we do something like, ignore the testIsolation setting (it's always legacy) unless experimental is enabled?

###### Legacy Mode

Cypress supports the `legacy` mode by default to ensure tests run in a
semi-clean browser context. When enabled Cypress will:

- [clears cookies](/api/commands/clearCookies) in current the domain
- [clears local storage](/api/commands/clearLocalStorage) in current the domain

The current test isolation implementation is named `legacy` because it has some
gaps which have led to leaked browser state between tests due to the page not

> clearing, cookies and local storage from other domains unintentionally

Reviewer: ```suggestion
clearing, resulting in cookies and local storage from other domains unintentionally
```

options.

###### On Mode

<Alert type="warning">

<strong class="alert-header"><Icon name="exclamation-triangle"></Icon>
Experimental</strong>

`on` mode is currently experimental and can be enabled by setting
the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config. This is the default test isolation behavior
when using the `experimentalSessionAndOrigin` experiment.

</Alert>

When in `on` mode, Cypress resets the browser context _before_ each test by:

- clearing the dom state by visiting `about:blank`
- clearing [cookies](/api/cypress-api/cookies) in all domains
- clearing
  [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage)
  in all domains
- clearing
  [`sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)
  in all domains

Because the test starts in a fresh browser context, you must re-visit your
application and perform the series of interactions needed to build the dom and
browser state for each test.

Additionally, the `cy.session()` command will inherent this mode and will clear

> the page and current browser context when establishing a browser session. so the

Reviewer: ```suggestion
the page and current browser context when establishing a browser session. This is so
```

tests can reliably pass when run standalone or in a randomized order.

###### Off Mode

<Alert type="warning">

<strong class="alert-header"><Icon name="exclamation-triangle"></Icon>
Experimental</strong>

`off` mode is currently experimental and can be enabled by setting
the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config.

</Alert>

> When in `off` mode, Cypress will not later the browser context before the test

Reviewer: ```suggestion
When in `off` mode, Cypress will not alter the browser context before the test
```

> starts. The page does not clear and cookies, local storage and session storage

Reviewer: ```suggestion
starts. The page does not clear between tests and cookies, local storage and session storage
```

| `legacy` (deprecated) | <br>- clears cookies in current domains<br>- local storage in current domains                                                                    | not available                                                                                                                                    |
| `on` (experimental)   | - clears page by visiting `about:blank`<br>- clears cookies in all domains<br>- local storage in all domains<br>- session storage in all domains | - clears page by visiting `about:blank`<br>- clears cookies in all domains<br>- local storage in all domains<br>- session storage in all domains |
| `off` (experimental)  | does not alter the current browser context                                                                                                       | <br>- clears cookies in all domains<br>- local storage in all domains<br>- session storage in all domains                                        |

##### Component testing

Cypress only support testIsolation `on` in component testing.

> When running component tests, the browser context will allow start in a clean

Reviewer: Ah I think the parity is starting to click for me now!

slate because Cypress will

- clear the page

> - clears [cookies](/api/cypress-api/cookies)

Reviewer 3: ```suggestion
- clear [cookies](/api/cypress-api/cookies)
```

> - clears

Reviewer 3: ```suggestion
- clear
```
